File: go/cmd/giztoy/commands/minimax/speech.go
package minimax

import (
	"bytes"
	"context"
	"fmt"
	"time"

	"github.com/spf13/cobra"

	"github.com/haivivi/giztoy/go/pkg/minimax"
)

var speechCmd = &cobra.Command{
	Use:   "speech",
	Short: "Speech synthesis service",
	Long: `Speech synthesis (TTS) service.

Supports synchronous, streaming, and asynchronous synthesis.

Example request file (speech.yaml):
  model: speech-2.6-hd
  text: Hello, this is a test message.
  voice_setting:
    voice_id: female-shaonv
    speed: 1.0
    vol: 1.0
    emotion: happy
  audio_setting:
    format: mp3
    sample_rate: 32000
  language_boost: Chinese`,
}

var speechSynthesizeCmd = &cobra.Command{
	Use:   "synthesize",
	Short: "Synthesize speech from text",
	Long: `Synthesize speech from text (synchronous).

Examples:
  giztoy minimax speech synthesize -f speech.yaml -o output.mp3
  giztoy minimax speech synthesize -f speech.yaml --json`,
	RunE: func(cmd *cobra.Command, args []string) error {
		if err := requireInputFile(); err != nil {
			return err
		}

		svc, err := loadServiceConfig()
		if err != nil {
			return err
		}

		var req minimax.SpeechRequest
		if err := loadRequest(inputFile, &req); err != nil {
			return err
		}

		if req.Model == "" {
			req.Model = minimax.ModelSpeech26HD
		}
		if req.VoiceSetting != nil && req.VoiceSetting.VoiceID == "" && svc.DefaultVoice != "" {
			req.VoiceSetting.VoiceID = svc.DefaultVoice
		}

		printVerbose("Model: %s", req.Model)
		printVerbose("Text length: %d characters", len(req.Text))

		client, err := createClientWith(svc)
		if err != nil {
			return err
		}

		reqCtx, cancel := context.WithTimeout(context.Background(), 120*time.Second)
		defer cancel()

		resp, err := client.Speech.Synthesize(reqCtx, &req)
		if err != nil {
			return fmt.Errorf("speech synthesis failed: %w", err)
		}

		if outputFile != "" && len(resp.Audio) > 0 {
			if err := outputBytes(resp.Audio, outputFile); err != nil {
				return fmt.Errorf("failed to write audio: %w", err)
			}
			printVerbose("Audio saved to: %s", outputFile)
		}

		result := map[string]any{
			"audio_size":  len(resp.Audio),
			"audio_url":   resp.AudioURL,
			"trace_id":    resp.TraceID,
			"extra_info":  resp.ExtraInfo,
			"output_file": outputFile,
		}
		return outputResult(result, "", outputJSON)
	},
}

var speechStreamCmd = &cobra.Command{
	Use:   "stream",
	Short: "Stream speech synthesis",
	Long: `Stream speech synthesis.

Audio will be streamed and saved incrementally.

Examples:
  giztoy minimax speech stream -f speech.yaml -o output.mp3`,
	RunE: func(cmd *cobra.Command, args []string) error {
		if err := requireInputFile(); err != nil {
			return err
		}
		if outputFile == "" {
			return fmt.Errorf("output file is required for streaming audio, use -o flag")
		}

		var req minimax.SpeechRequest
		if err := loadRequest(inputFile, &req); err != nil {
			return err
		}
		if req.Model == "" {
			req.Model = minimax.ModelSpeech26HD
		}

		printVerbose("Streaming to: %s", outputFile)

		client, err := createClient()
		if err != nil {
			return err
		}

		reqCtx, cancel := context.WithTimeout(context.Background(), 300*time.Second)
		defer cancel()

		var audioBuf bytes.Buffer
		var lastChunk *minimax.SpeechChunk

		for chunk, err := range client.Speech.SynthesizeStream(reqCtx, &req) {
			if err != nil {
				return fmt.Errorf("streaming failed: %w", err)
			}
			if chunk.Audio != nil {
				audioBuf.Write(chunk.Audio)
			}
			lastChunk = chunk
		}

		if err := outputBytes(audioBuf.Bytes(), outputFile); err != nil {
			return fmt.Errorf("failed to write audio: %w", err)
		}
		printSuccess("Audio saved to: %s (%s)", outputFile, formatBytes(audioBuf.Len()))

		if lastChunk != nil && lastChunk.ExtraInfo != nil {
			result := map[string]any{
				"audio_size":  audioBuf.Len(),
				"extra_info":  lastChunk.ExtraInfo,
				"trace_id":    lastChunk.TraceID,
				"output_file": outputFile,
			}
			return outputResult(result, "", outputJSON)
		}
		return nil
	},
}

var speechAsyncCmd = &cobra.Command{
	Use:   "async",
	Short: "Create async speech synthesis task",
	Long: `Create an asynchronous speech synthesis task for long text.

Supports up to 1,000,000 characters. Returns a task ID for tracking.

Examples:
  giztoy minimax speech async -f long-text.yaml
  giztoy minimax speech async -f long-text.yaml --json`,
	RunE: func(cmd *cobra.Command, args []string) error {
		if err := requireInputFile(); err != nil {
			return err
		}

		var req minimax.AsyncSpeechRequest
		if err := loadRequest(inputFile, &req); err != nil {
			return err
		}
		if req.Model == "" {
			req.Model = minimax.ModelSpeech26HD
		}

		printVerbose("Text length: %d characters", len(req.Text))

		client, err := createClient()
		if err != nil {
			return err
		}

		reqCtx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
		defer cancel()

		task, err := client.Speech.CreateAsyncTask(reqCtx, &req)
		if err != nil {
			return fmt.Errorf("create async task failed: %w", err)
		}

		printSuccess("Async task created: %s", task.ID)

		result := map[string]any{
			"task_id": task.ID,
			"status":  "created",
		}
		return outputResult(result, outputFile, outputJSON)
	},
}

func init() {
	speechCmd.AddCommand(speechSynthesizeCmd)
	speechCmd.AddCommand(speechStreamCmd)
	speechCmd.AddCommand(speechAsyncCmd)
}

File: examples/rust/minimax/src/bin/speech.rs
//! Speech synthesis example.
//!
//! This example demonstrates how to use the MiniMax SDK for speech synthesis.
//!
//! Run with:
//! ```bash
//! export MINIMAX_API_KEY="your-api-key"
//! cargo run --example minimax-speech
//! ```

use std::env;
use std::pin::pin;

use giztoy_minimax::{
    AudioSetting, Client, SpeechRequest, VoiceSetting,
    MODEL_SPEECH_26_HD, VOICE_FEMALE_SHAONV,
};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Get API key from environment
    let api_key = env::var("MINIMAX_API_KEY")
        .expect("MINIMAX_API_KEY environment variable not set");

    // Create client
    let client = Client::new(api_key)?;

    // Example 1: Simple speech synthesis
    println!("Example 1: Simple speech synthesis");
    
    let request = SpeechRequest {
        model: MODEL_SPEECH_26_HD.to_string(),
        text: "你好，这是一个语音合成的示例。Hello, this is a speech synthesis example.".to_string(),
        voice_setting: Some(VoiceSetting {
            voice_id: VOICE_FEMALE_SHAONV.to_string(),
            speed: Some(1.0),
            ..Default::default()
        }),
        audio_setting: Some(AudioSetting {
            sample_rate: Some(32000),
            ..Default::default()
        }),
        ..Default::default()
    };

    let response = client.speech().synthesize(&request).await?;
    
    // Save audio to file
    std::fs::write("output_simple.mp3", &response.audio)?;
    println!("Audio saved to output_simple.mp3 ({} bytes)", response.audio.len());
    
    if let Some(info) = &response.extra_info {
        println!("Duration: {} ms", info.audio_length);
        println!("Sample rate: {}", info.audio_sample_rate);
    }

    // Example 2: Streaming speech synthesis
    println!("\nExample 2: Streaming speech synthesis");
    
    use futures::StreamExt;
    
    let stream_request = SpeechRequest {
        model: MODEL_SPEECH_26_HD.to_string(),
        text: "这是一个流式语音合成的示例，音频会边生成边返回。".to_string(),
        voice_setting: Some(VoiceSetting {
            voice_id: VOICE_FEMALE_SHAONV.to_string(),
            ..Default::default()
        }),
        ..Default::default()
    };

    let mut audio_data = Vec::new();
    let speech_client = client.speech();
    let stream = speech_client.synthesize_stream(&stream_request).await?;
    let mut stream = pin!(stream);
    
    let mut chunk_count = 0;
    while let Some(chunk) = stream.next().await {
        let chunk = chunk?;
        audio_data.extend_from_slice(&chunk.audio);
        chunk_count += 1;
        print!(".");
    }
    println!();
    
    std::fs::write("output_stream.mp3", &audio_data)?;
    println!("Audio saved to output_stream.mp3 ({} bytes, {} chunks)", audio_data.len(), chunk_count);

    println!("\nDone!");
    Ok(())
}
